bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/LegalEntityService.kt

```
@@ -79,71 +76,43 @@ class LegalEntityService(
-            val sharingStatusType = sharingStatus?.status
-
-            if (sharingStatusType == null || sharingStatusType.isError()) {
-                // ERROR: SharingProcessError
-                errors.add(
-                    outputSaasMappingService.buildErrorInfoSharingProcessError(externalId, sharingStatus))
-            } else if (bpn != null) {
-                val legalEntity = legalEntitiesByBpnMap[bpn]
-                val legalAddress = legalAddressesByBpnMap[bpn]
-                if (legalEntity != null && legalAddress != null) {
-                    // OKAY: entry found in pool
-                    validContent.add(
-                        toLegalEntityOutput(externalId, legalEntity, legalAddress)
-                    )
-                } else {
-                    // ERROR: BpnNotInPool
-                    errors.add(
-                        outputSaasMappingService.buildErrorInfoBpnNotInPool(externalId, bpn))
-                }
-            } else if (outputSaasMappingService.isSharingTimeoutReached(partner)) {
-                // ERROR: SharingTimeout
-                errors.add(
-                    outputSaasMappingService.buildErrorInfoSharingTimeout(externalId, partner.lastModifiedAt)
-                )
-            } else {
-                pendingExternalIds.add(externalId)
-            }
+        val partnersWithPoolBpn = partnersWithLocalBpn.filter { legalEntitiesByBpnMap[it.bpn] != null }
+        val bpnByExternalIdMap = partnersWithPoolBpn.associate { Pair(it.externalId, it.bpn) }
+
+        //Evaluate the sharing status of the legal entities
+        val sharingStatus = outputSaasMappingService.evaluateSharingStatus(partners, partnersWithLocalBpn, partnersWithPoolBpn)
+
+        val validLegalEntities = sharingStatus.validExternalIds.map { externalId ->
+            val bpn = bpnByExternalIdMap[externalId]!!
+            val legalEntity = legalEntitiesByBpnMap[bpn]!!
+            val legalAddress = legalAddressesByBpnMap[bpn]!!
```

We made sure that for all the `validExternalIds` there is a legal entity, but we theoretically don't know that there is a corresponding legal address. 
This might be just a hypothetical case but maybe we could at least throw a meaningful exception?
The same goes for sites and their main addresses.

---

made it behave like before, when a legal address or main address is missing, the entry is added to the BpnNotInPool Error status